### ISSUE TITLE
`--runemptysuite` should work with `--rerunfailed` when there are no failed tests

### ISSUE THREAD

**Issue by villesau** [robotframework/robotframework#3862]
Hi,

The `Collecting failed tests from 'output.xml' failed: All tests passed.` is pretty misleading: In reality nothing fails, it finds as just many failed tests as it is supposed to find. Maybe `Could not find failed tests, exiting` or similar would be better. Now it indicates that it would give some non-zero status code and so on.

**Comment by pekkaklarck** [comment]
I'm personally fine with the current error message but don't mind it being changed if someone provides a PR. Are you @villesau interested? Notice that you get a non-zero status code (252) in this case.

**Comment by villesau** [comment]
So it _is_ an error of some sort? Now it gets even more confusing 🙈 How can I do it _without_ erroring? I find all tests passing very much opposite to any erroneous case.

**Comment by pekkaklarck** [comment]
Yes, if you explicitly tell Robot to run failed tests but there are no failures that's considered an error. How should it work in your opinion?

**Comment by pekkaklarck** [comment]
In practice this is considered similar situation than using e.g. `--include` so that no test matches. In that usage you can use `--runemptysuite` to tell Robot to execute an empty suite instead of failing. We certainly could enhance `--runemptysuite` so that it would work also with `--rerunfailed`.

**Comment by t-my** [comment]
Just return a zero (success) exit code if `--rerunfailed` is given and there are no failed tests to run. Currently you need to do some amount of scripting to support a ci/cd job where you 1. run a suite, 2. run any failed tests.

**Comment by pekkaklarck** [comment]
In my opinion `--rerunfailed` failing if there's nothing to run by default is fine. What could, and arguably should, be changed is that `--runemptysuite`could be used change the behavior.

**Comment by ChrissNCS** [comment]
Ha! I did not think that someone else run into the same problem ;D
We are also using a ci/cd pipeline where we want to run tests in the first job and rerun failed tests in the second one after.

In case of gitlab, you want to throw a warning for the first job (error would set the whole pipeline as failed) and deliver the artifacts (test report) to the second job. In a perfekt world, I would say the job should only be executed if the first job thrown an warning but gitlab does not provide this... (on_failure is possibile but than the story with the whole pipeline..)
So we need to run the second one in every case...

Long story short: I would be really fine with the combination of  '--runemptysuite --rerunfailed'

**Comment by t-my** [comment]
To be a bit more specific about requirements: the return code of the program should be a success (zero) if there are no tests to run. Currently it returns a non-zero return code. 

This will be helpful in automated pipelines where you firstly run all tests and then rerun any failed tests.

**Comment by pekkaklarck** [comment]
I don't think it's a good idea to change the default behavior if there are no tests to run. Getting a zero return code when, for example, executing an empty directory by accident doesn't sound good to me. Those who want that behavior can already now use `--runemptysuite` in their start-up scripts to make it the default behavior. It would be even more convenient if we had support for configuration files where this kind of options could be configured on system or user level.

This issue is about `--runemptysuite` not being honored with `--rerunfailed`. That' definitely a missing feature or even a bug. I'm starting to lean considering it a bug which would mean this could be fixed already in RF 5.0.1.

**Comment by t-my** [comment]
Oh sorry, I wasn't articulating clearly. We can assume that changing  the default behaviour would be a breaking change. 

I meant to say that should return success when run with parameter `--runemptysuite` AND `--rerunfailed` AND there is no tests to be run.

**Comment by pekkaklarck** [comment]
Fixing this was a bit more challenging than I thought due to how `--rerunfailed` has been implemented. Now I got a pretty nice fix done but still need to create acceptance tests for it.

The fix will contain a subtle change to `TestSuite.filter` behavior. This method basically implements `--include/--exclude` and `--suite/--test` behavior and it accepts lists of include/exclude tags and suite/test names as arguments. Earlier passing, for example, test names as `None` or an empty list caused no tests to be filtered. After the forthcoming change `None` still works the same way, but passing an empty list means no test will be selected. That's obviously a backwards incompatible change, but because the default behavior doesn't change and because using `None` is anyway better if you want to explicitly disable filtering using a certain criteria, I don't think this change needs to be mentioned in the release notes.

**Comment by t-my** [comment]
I can try to test this, is there a pull request etc?

**Comment by pekkaklarck** [comment]
Code has been pushed to the master branch. You can either install the latest code from there or wait for the release candidate that I plan to create later this week, probably Thursday evening.
